In [1]:
import pandas as pd
df = pd.read_csv(r'F:\Harshi College\MS DA\Machine Learning\lab\unsupervised learning\weather.csv',sep=',')
print(df.head)
print(df.columns.values)
# df = df.reset_index()

<bound method NDFrame.head of      MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
0        8.0     24.3       0.0          3.4       6.3          NW   
1       14.0     26.9       3.6          4.4       9.7         ENE   
2       13.7     23.4       3.6          5.8       3.3          NW   
3       13.3     15.5      39.8          7.2       9.1          NW   
4        7.6     16.1       2.8          5.6      10.6         SSE   
..       ...      ...       ...          ...       ...         ...   
361      9.0     30.7       0.0          7.6      12.1         NNW   
362      7.1     28.4       0.0         11.6      12.7           N   
363     12.5     19.9       0.0          8.4       5.3         ESE   
364     12.5     26.9       0.0          5.0       7.1          NW   
365     12.3     30.2       0.0          6.0      12.6          NW   

     WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  ...  Humidity3pm  \
0             30.0         SW         NW           6.

#####  EDA

In [86]:
# DataFrame. describe(include = 'all') to get a summary of all the columns when the dataframe has mixed column types. The default behavior is to only provide a summary for the numerical columns
print(df.describe(include = "all"))

           MinTemp     MaxTemp    Rainfall  Evaporation    Sunshine  \
count   366.000000  366.000000  366.000000   366.000000  363.000000   
unique         NaN         NaN         NaN          NaN         NaN   
top            NaN         NaN         NaN          NaN         NaN   
freq           NaN         NaN         NaN          NaN         NaN   
mean      7.265574   20.550273    1.428415     4.521858    7.909366   
std       6.025800    6.690516    4.225800     2.669383    3.481517   
min      -5.300000    7.600000    0.000000     0.200000    0.000000   
25%       2.300000   15.025000    0.000000     2.200000    5.950000   
50%       7.450000   19.650000    0.000000     4.200000    8.600000   
75%      12.500000   25.500000    0.200000     6.400000   10.500000   
max      20.900000   35.800000   39.800000    13.800000   13.600000   

       WindGustDir  WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  ...  \
count          363     364.000000        335        365    359.000000

In [87]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MinTemp        366 non-null    float64
 1   MaxTemp        366 non-null    float64
 2   Rainfall       366 non-null    float64
 3   Evaporation    366 non-null    float64
 4   Sunshine       363 non-null    float64
 5   WindGustDir    363 non-null    object 
 6   WindGustSpeed  364 non-null    float64
 7   WindDir9am     335 non-null    object 
 8   WindDir3pm     365 non-null    object 
 9   WindSpeed9am   359 non-null    float64
 10  WindSpeed3pm   366 non-null    int64  
 11  Humidity9am    366 non-null    int64  
 12  Humidity3pm    366 non-null    int64  
 13  Pressure9am    366 non-null    float64
 14  Pressure3pm    366 non-null    float64
 15  Cloud9am       366 non-null    int64  
 16  Cloud3pm       366 non-null    int64  
 17  Temp9am        366 non-null    float64
 18  Temp3pm   

In [88]:
# to find all the columns with the NaN values
nan_values = df.isna().any()
print(nan_values)

# to select all those columns with NaNs
nan_values = df[df.columns[df.isna().any()]]
print(nan_values)

MinTemp          False
MaxTemp          False
Rainfall         False
Evaporation      False
Sunshine          True
WindGustDir       True
WindGustSpeed     True
WindDir9am        True
WindDir3pm        True
WindSpeed9am      True
WindSpeed3pm     False
Humidity9am      False
Humidity3pm      False
Pressure9am      False
Pressure3pm      False
Cloud9am         False
Cloud3pm         False
Temp9am          False
Temp3pm          False
RainToday        False
RISK_MM          False
RainTomorrow     False
dtype: bool
     Sunshine WindGustDir  WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am
0         6.3          NW           30.0         SW         NW           6.0
1         9.7         ENE           39.0          E          W           4.0
2         3.3          NW           85.0          N        NNE           6.0
3         9.1          NW           54.0        WNW          W          30.0
4        10.6         SSE           50.0        SSE        ESE          20.0
..        ...       

In [91]:
# # checking for NaNs and infs
# import numpy as np
# print(np.any(np.isnan(df)))
# print(np.all(np.isfinite(df)))

In [92]:
# to get Data types of columns in Dataframe
print(df. dtypes)

MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm       int64
Humidity9am        int64
Humidity3pm        int64
Pressure9am      float64
Pressure3pm      float64
Cloud9am           int64
Cloud3pm           int64
Temp9am          float64
Temp3pm          float64
RainToday         object
RISK_MM          float64
RainTomorrow      object
dtype: object


##### data preprocessing

In [93]:
# to get Data types of columns in Dataframe
print(df. dtypes)

# converting all non-numeric columns to numeric columns

# enc.fit() creates the corresponding integer values.
# enc.transform() applies the encoding to the df values.
from sklearn.preprocessing import LabelEncoder
enc = LabelEncoder()

# Specify the dtype to extract with the include parameter.
non_numeric_data = df.select_dtypes(include=object)
non_numeric_columns = non_numeric_data.columns.values
print("\nnon_numeric_columns in the dataframe are as follows :\n", non_numeric_columns)
for i in range(len(non_numeric_columns)):
    col = non_numeric_columns[i]
#     print(df[col])
    
    enc.fit(df[col])
    df[col] = enc.transform(df[col])
#     print(df[col])

# to ensure Data types of columns in Dataframe is numeric
print(df. dtypes)

MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm       int64
Humidity9am        int64
Humidity3pm        int64
Pressure9am      float64
Pressure3pm      float64
Cloud9am           int64
Cloud3pm           int64
Temp9am          float64
Temp3pm          float64
RainToday         object
RISK_MM          float64
RainTomorrow      object
dtype: object

non_numeric_columns in the dataframe are as follows :
 ['WindGustDir' 'WindDir9am' 'WindDir3pm' 'RainToday' 'RainTomorrow']
MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir        int32
WindGustSpeed    float64
WindDir9am         int32
WindDir3pm         int32
WindSpeed9am     float64
WindSpeed3pm       int64
Humidity9am        int64
Humidity3pm

In [94]:
df['RainTomorrow'] = df['RainTomorrow'].replace(['Yes'],1)
df['RainTomorrow'] = df['RainTomorrow'].replace(['No'],0)
df['RainTomorrow']

0      1
1      1
2      1
3      1
4      0
      ..
361    0
362    0
363    0
364    0
365    0
Name: RainTomorrow, Length: 366, dtype: int32

##### train-test split

In [95]:
from sklearn.model_selection import train_test_split

X = df.drop(['RainTomorrow'], axis = 1)
y = df['RainTomorrow']

training_data, testing_data, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=5)

print(f"No. of training examples: {training_data.shape[0]}")
print(f"No. of testing examples: {testing_data.shape[0]}")

No. of training examples: 292
No. of testing examples: 74


In [96]:
# ValueError: Input contains NaN, infinity or a value too large for dtype('float64').. the below two blocks to handle this error
# https://stackabuse.com/k-nearest-neighbors-algorithm-in-python-and-scikit-learn/
# https://researchdatapod.com/python-valueerror-input-contains-nan-infinity-or-a-value-too-large-for-dtype-float64/

In [97]:
# Feature Scaling -  to scale the features so that all of them can be uniformly evaluated.

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(training_data)

training_data = scaler.transform(training_data)
testing_data = scaler.transform(testing_data)
training_data = pd.DataFrame(training_data)
testing_data = pd.DataFrame(testing_data)

In [98]:
from sklearn.impute import SimpleImputer
imp_mean = SimpleImputer(missing_values=np.nan, strategy='constant', fill_value=0)
imputer = imp_mean.fit(training_data)
training_data = imputer.transform(training_data)

imp_mean = SimpleImputer(missing_values=np.nan, strategy='constant', fill_value=0)
imputer = imp_mean.fit(testing_data)
testing_data = imputer.transform(testing_data)

##### KNN classifier

In [99]:
# KNN classifier

from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)

accList = []
# Import necessary modules
from sklearn.neighbors import KNeighborsClassifier
for i in range(40,45):
    knn = KNeighborsClassifier(n_neighbors=i)

    knn.fit(training_data, y_train)

    # Predict on dataset which model has not seen before
    y_pred = knn.predict(testing_data)
#     print(y_pred)

    import numpy as np
    # converting series to ndarray
    y_test = np.asarray(y_test)

    # confusion matrix
    from sklearn.metrics import confusion_matrix
    tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#     print(confusion_matrix(y_test, y_pred))
    print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

    # error evaluation methods
    # calculating accuracy

    from sklearn.metrics import classification_report
    from sklearn.metrics import accuracy_score

    #Check precision, recall, f1-score
    print(classification_report(y_test, y_pred) )

    print("accuracy_score",accuracy_score(y_test, y_pred))
    print("is with k = ",i)
    accList.append(accuracy_score(y_test, y_pred))
    print("=======================================================================================")

from statistics import mean
print("mean of the accuracy is")
mean(accList)


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:35:40
true positive : 58
 true negative : 0
 false positive : 14
 false negative : 2

              precision    recall  f1-score   support

           0       0.81      1.00      0.89        58
           1       1.00      0.12      0.22        16

    accuracy                           0.81        74
   macro avg       0.90      0.56      0.56        74
weighted avg       0.85      0.81      0.75        74

accuracy_score 0.8108108108108109
is with k =  40
true positive : 58
 true negative : 0
 false positive : 14
 false negative : 2

              precision    recall  f1-score   support

           0       0.81      1.00      0.89        58
           1       1.00      0.12      0.22        16

    accuracy                           0.81        74
   macro avg       0.90      0.56      0.56        74
weighted avg       0.85      0.81      0.75        74

accuracy_score 0.8108108108108109
is with k =  41
true positive : 58
 true negative : 0
 false positive : 14
 fals

##### Decision tree 

In [100]:
# Decision tree
from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)


from sklearn.tree import DecisionTreeClassifier
clf_model = DecisionTreeClassifier(criterion="gini", random_state=42, max_depth=3, min_samples_leaf=5)   
clf_model.fit(training_data,y_train)

y_pred = clf_model.predict(testing_data)
y_pred

# confusion matrix
from sklearn.metrics import confusion_matrix
tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#print(confusion_matrix(y_test, y_pred))
print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

# error evaluation methods
# calculating accuracy

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

#Check precision, recall, f1-score
print(classification_report(y_test, y_pred) )

print("accuracy_score",accuracy_score(y_test, y_pred))


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:35:40
true positive : 58
 true negative : 0
 false positive : 0
 false negative : 16

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        58
           1       1.00      1.00      1.00        16

    accuracy                           1.00        74
   macro avg       1.00      1.00      1.00        74
weighted avg       1.00      1.00      1.00        74

accuracy_score 1.0
end time = 23:35:40
0:00:00


##### Naive Bayes classifier

In [101]:
# ValueError: Negative values in data passed to MultinomialNB (input X). 
#     MinMaxScaler i sused to handle the above error
# https://stackoverflow.com/questions/24169238/dealing-with-negative-values-in-sklearn-multinomialnb

In [102]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
print(scaler.fit(training_data))
# print(scaler.training_data_max_)
training_data = (scaler.transform(training_data))
training_data = pd.DataFrame(training_data)

testing_data = scaler.transform(testing_data)
testing_data = pd.DataFrame(testing_data)

MinMaxScaler()


In [103]:
# Naïve Bayes classifier
from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)

from sklearn.naive_bayes import MultinomialNB

naive_bayes_classifier = MultinomialNB()
naive_bayes_classifier.fit(training_data, y_train)

y_pred = naive_bayes_classifier.predict(testing_data)
len(y_pred)

# confusion matrix
from sklearn.metrics import confusion_matrix
tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#print(confusion_matrix(y_test, y_pred))
print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

# error evaluation methods
# calculating accuracy

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

#Check precision, recall, f1-score
print(classification_report(y_test, y_pred) )

print("accuracy_score",accuracy_score(y_test, y_pred))


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:35:40
true positive : 58
 true negative : 0
 false positive : 13
 false negative : 3

              precision    recall  f1-score   support

           0       0.82      1.00      0.90        58
           1       1.00      0.19      0.32        16

    accuracy                           0.82        74
   macro avg       0.91      0.59      0.61        74
weighted avg       0.86      0.82      0.77        74

accuracy_score 0.8243243243243243
end time = 23:35:40
0:00:00


##### logistic regression

In [104]:
# Logistic regression

from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)


# Fitting Logistic Regression to the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0, solver='lbfgs', multi_class='auto')
classifier.fit(training_data, y_train)

# Predicting the Test set results
y_pred = classifier.predict(testing_data)
y_pred

# confusion matrix
from sklearn.metrics import confusion_matrix
tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#print(confusion_matrix(y_test, y_pred))
print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

# error evaluation methods
# calculating accuracy

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

#Check precision, recall, f1-score
print(classification_report(y_test, y_pred) )

print("accuracy_score",accuracy_score(y_test, y_pred))


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:35:40
true positive : 57
 true negative : 1
 false positive : 8
 false negative : 8

              precision    recall  f1-score   support

           0       0.88      0.98      0.93        58
           1       0.89      0.50      0.64        16

    accuracy                           0.88        74
   macro avg       0.88      0.74      0.78        74
weighted avg       0.88      0.88      0.86        74

accuracy_score 0.8783783783783784
end time = 23:35:40
0:00:00


##### summary

accuracy score for weather dataset analysis:

KNN classifier - 0.8108108108108109

decision tree - 1.0

Naive bayes classifier - 0.8783783783783784

Logistic regression - 0.918918918918919